File: python/examples/jobs/02_job_management_sdk.py
#!/usr/bin/env python3
"""
Example: Job management using SDK

This example demonstrates:
- Getting job details
- Cancelling running jobs
- Deleting job results to free storage
- Getting presigned download URLs
"""

import asyncio
from crawl4ai_cloud import AsyncWebCrawler

# Configuration
API_KEY = "YOUR_API_KEY"  # Replace with your API key


async def main():
    async with AsyncWebCrawler(api_key=API_KEY) as crawler:
        # Create an async job for testing
        print("=== Creating Test Job ===")
        results = await crawler.run_many(
            urls=["https://example.com", "https://example.org"],
            wait=False  # Don't wait, just create the job
        )
        # When wait=False, we get a job object back
        print(f"Created job")

        # Get job details
        print("\n=== Get Job Details ===")
        jobs = await crawler.list_jobs(limit=1)
        if jobs.jobs:
            job = jobs.jobs[0]
            print(f"Job ID: {job.job_id}")
            print(f"Status: {job.status}")
            print(f"URLs: {job.urls}")
            print(f"Created: {job.created_at}")

            # Cancel the job (without deleting results)
            print("\n=== Cancel Job (Keep Results) ===")
            cancelled = await crawler.cancel_job(job.job_id, delete_results=False)
            print(f"Status: {cancelled.status}")

        # Create another job and delete it completely
        print("\n=== Cancel + Delete Results ===")
        results2 = await crawler.run_many(
            urls=["https://example.com"],
            wait=False
        )

        jobs2 = await crawler.list_jobs(limit=1)
        if jobs2.jobs:
            job2 = jobs2.jobs[0]
            print(f"Created job: {job2.job_id}")
            deleted = await crawler.cancel_job(job2.job_id, delete_results=True)
            print(f"Deleted: {deleted.status}")

        # Get download URL for completed job (example)
        print("\n=== Get Download URL ===")
        try:
            completed_jobs = await crawler.list_jobs(status="completed", limit=1)
            if completed_jobs.jobs:
                job_id = completed_jobs.jobs[0].job_id
                download_url = await crawler.get_download_url(job_id, expires_in=3600)
                print(f"Download URL: {download_url[:100]}...")
                print("URL expires in 3600 seconds (1 hour)")
            else:
                print("No completed jobs found")
        except Exception as e:
            print(f"Error: {e}")


if __name__ == "__main__":
    asyncio.run(main())

File: nodejs/examples/jobs/02_job_management_sdk.ts
#!/usr/bin/env npx ts-node
/**
 * Example: Job management using SDK
 *
 * This example demonstrates:
 * - Getting job details
 * - Cancelling running jobs
 * - Getting presigned download URLs
 *
 * Usage:
 *     npx ts-node 02_job_management_sdk.ts
 */

import { AsyncWebCrawler, CrawlJob } from 'crawl4ai-cloud';

// Configuration
const API_KEY = 'YOUR_API_KEY'; // Replace with your API key

async function main(): Promise<void> {
  const crawler = new AsyncWebCrawler({ apiKey: API_KEY });

  try {
    // Create an async job for testing
    console.log('=== Creating Test Job ===');
    const result = (await crawler.runMany(['https://example.com', 'https://example.org'], {
      wait: false, // Don't wait, just create the job
    })) as CrawlJob;

    console.log(`Created job: ${result.id}`);

    // Get job details
    console.log('\n=== Get Job Details ===');
    const job = await crawler.getJob(result.id, false);
    console.log(`Job ID: ${job.id}`);
    console.log(`Status: ${job.status}`);
    console.log(`URLs: ${job.urlsCount}`);
    console.log(`Created: ${job.createdAt}`);

    // Wait for job and get results
    console.log('\n=== Wait for Job ===');
    const completedJob = await crawler.waitJob(result.id, {
      pollInterval: 2,
      includeResults: true,
    });
    console.log(`Final Status: ${completedJob.status}`);
    console.log(`Progress: ${completedJob.progress.completed}/${completedJob.progress.total}`);

    // Cancel a job (create another one first)
    console.log('\n=== Cancel Job ===');
    const job2 = (await crawler.runMany(['https://example.com'], {
      wait: false,
    })) as CrawlJob;
    console.log(`Created job: ${job2.id}`);

    const cancelled = await crawler.cancelJob(job2.id);
    console.log(`Cancelled: ${cancelled}`);

    // Get download URL for completed job
    console.log('\n=== Get Download URL ===');
    try {
      const completedJobs = await crawler.listJobs({ status: 'completed', limit: 1 });
      if (completedJobs.length > 0) {
        const jobId = completedJobs[0].id;
        const downloadUrl = await crawler.downloadUrl(jobId, 3600);
        console.log(`Download URL: ${downloadUrl.slice(0, 100)}...`);
        console.log('URL expires in 3600 seconds (1 hour)');
      } else {
        console.log('No completed jobs found');
      }
    } catch (e) {
      console.log(`Error: ${e}`);
    }
  } finally {
    await crawler.close();
  }
}

main().catch(console.error);
